### PR TITLE
Refactor Writer class

### PR DESCRIPTION
**Pull request by Rodrigo-Tenorio**
Implement the argument-independent first introduced in #39 as part of BinaryModulatedWriter into Writer.

### CODE REVIEW

File: pyfstat/make_sfts.py [hunk 1]
```diff
@@ -272,157 +293,29 @@ def basic_setup(self):
...
-                window,
-                transientStartTime,
-                transientTau,
-            )
-
-    def make_cff(self, verbose=False):
-        """
-        Generates a .cff file
-
+    def parse_args_consistent_with_mfd(self):
```
Comment (Rodrigo-Tenorio) [inline pyfstat/make_sfts.py:296]: Here's where all the magic happens: This method looks into the class itself using the `__dict__` attribute. If any of the attributes specified in `signal_parameter_labels` are not `None`, they are included into any output of
`get_single_config_line`, which is the main way for MFD to get signal parameters.

This way, `BinaryModulatedWriter` only needs to include its own parameters and nothing more.

**Comment by Rodrigo-Tenorio** [comment]
This change works fine for `Writer` and `BinaryModulatedWriter`.

The problem now is `GlitchWriter` and the way in which `get_single_config_line` is used in there.

**Comment by dbkeitel** [comment]
> The problem now is `GlitchWriter` and the way in which `get_single_config_line` is used in there.

If you don't want to spend time dealing with that one right now, it's fine if you just move the get_single function(s) to be methods of that child class. That way we have the main `Writer` cleaned up and can revisit the glitchy one later.

**Comment by Rodrigo-Tenorio** [comment]
> > The problem now is `GlitchWriter` and the way in which `get_single_config_line` is used in there.
> 
> If you don't want to spend time dealing with that one right now, it's fine if you just move the get_single function(s) to be methods of that child class. That way we have the main `Writer` cleaned up and can revisit the glitchy one later.

I've just checked what would have to be done in order to make this work: 
So far I've been assuming each `Writer` injects one signal at a time; as obvious as it sounds, `GlitchWriter` injects more than one signal (i.e. a "glitch" is two signals with a discontinuity), so it's not going to be as trivial as for the binary class.

I'll take your advice.

**Comment by Rodrigo-Tenorio** [comment]
Looks like I messed up something in here. I'll look into this after implementing a test in #145 
PyFstat_example_semi_coherent_binary_search_using_MCMC_walkers

**Comment by Rodrigo-Tenorio** [comment]
> Looks like I messed up something in here. I'll look into this after implementing a test in #145

Solved in cf57a67, it was just a matter of class attributes not being properly set.
PyFstat_example_semi_coherent_binary_search_using_MCMC_walkers

File: pyfstat/make_sfts.py [hunk 2]
```diff
@@ -722,6 +657,15 @@ def __init__(
             [transientWindowType,transientStartTime,transientTau] instead!
         see `lalapps_Makefakedata_v5 --help` for help with the other paramaters
         """
+        self.signal_parameter_labels = super().signal_parameter_labels + [
```
Comment (dbkeitel) [inline pyfstat/make_sfts.py:660]: very elegant!

File: pyfstat/make_sfts.py [hunk 3]
```diff
@@ -484,6 +377,40 @@ def calculate_fmin_Band(self):
...
+            logging.info(
+                "Writing the following injection parameters"
+                " to config file {:s}:".format(self.config_file_name)
+            )
+            logging.info(content)
+
+        if self.check_if_cff_file_needs_rewritting(content):
+            config_file = open(self.config_file_name, "w+")
+            config_file.write(content)
+            config_file.close()
```
Comment (dbkeitel) [inline pyfstat/make_sfts.py:412]: seems we lost some verbosity option here, but it didn't seem very informative, so I guess that's ok

File: pyfstat/make_sfts.py [hunk 4]
```diff
@@ -938,6 +797,112 @@ def __init__(
             np.array([delta_phi, delta_F0, delta_F1, delta_F2]).T
         )
 
+    def get_base_template(self, i, Alpha, Delta, h0, cosi, psi, phi, F0, F1, F2, tref):
```
Comment (dbkeitel) [inline pyfstat/make_sfts.py:800]:
```suggestion
    def get_base_template(self, i, Alpha, Delta, h0, cosi, psi, phi, F0, F1, F2, tref):
        """ FIXME: ported over from Writer, should be replaced by a more elegant re-use of parse_args_consistent_with_mfd """
```
(black may require linebreaks or something)

File: pyfstat/make_sfts.py [hunk 5]
```diff
@@ -272,157 +293,29 @@ def basic_setup(self):
...
-                transientTau,
-            )
-
-    def make_cff(self, verbose=False):
-        """
-        Generates a .cff file
-
+    def parse_args_consistent_with_mfd(self):
+        """ This will allow us to get rid of the get_single_config_* family
+            Future rework may use a dictionary as the default input method,
```
Comment (dbkeitel) [inline pyfstat/make_sfts.py:298]: I think these two comments are now resolved. :+1:

File: pyfstat/make_sfts.py (hunk 1 above)
Comment (dbkeitel) [reply pyfstat/make_sfts.py:296]: Makes a lot of sense!

File: pyfstat/make_sfts.py (hunk 3 above)
Comment (Rodrigo-Tenorio) [reply pyfstat/make_sfts.py:412]: Ups, this was not on purpose, I'll add it: It's useful to have the actual parameters after parsing on display (*ejm*)

Comment (Rodrigo-Tenorio) [reply pyfstat/make_sfts.py:412]: Actually... we did not loose any verbosity, did we?

**Review by dbkeitel — APPROVED** [review]

**Pull request merged**